helios-core/src/main/kotlin/helios/instances/instancesEq.kt

```diff
@@ -0,0 +1,85 @@
+package helios.instances
+
+import arrow.core.extensions.eq
+import arrow.data.extensions.list.foldable.forAll
+import arrow.extension
+import arrow.typeclasses.Eq
+import helios.core.*
+import helios.instances.jsarray.eq.eq
+import helios.instances.jsnumber.eq.eq
+import helios.instances.jsobject.eq.eq
+import helios.instances.json.eq.eq
+
+
+@extension
+interface JsObjectEqInstance : Eq<JsObject> {
+  override fun JsObject.eqv(b: JsObject): Boolean = with(Json.eq()) {
+    this@eqv.value.entries.zip(b.value.entries) { aa, bb ->
+      aa.key == bb.key && aa.value.eqv(bb.value)
+    }.forAll { it }
+  }
+}
+
+@extension
+interface JsArrayEqInstance : Eq<JsArray> {
+  override fun JsArray.eqv(b: JsArray): Boolean = with(Json.eq()) {
+    this@eqv.value.zip(b.value) { a, b -> a.eqv(b) }
+      .forAll { it }
+  }
+}
+
+@extension
+interface JsonEqInstance : Eq<Json> {
+  override fun Json.eqv(b: Json): Boolean = when {
+    this is JsObject && b is JsObject -> JsObject.eq().run { this@eqv.eqv(b) }
+    this is JsString && b is JsString -> String.eq().run {
+      this@eqv.value.toString().eqv(b.value.toString())
+    }
+    this is JsNumber && b is JsNumber -> JsNumber.eq().run { this@eqv.eqv(b) }
+    this is JsBoolean && b is JsBoolean -> Boolean.eq().run { this@eqv.value.eqv(b.value) }
+    this is JsArray && b is JsArray -> JsArray.eq().run { this@eqv.eqv(b) }
+    else -> this.isNull && b.isNull
+  }
+
+}
+
+@extension
+interface JsNumberEqInstance : Eq<JsNumber> {
+  override fun JsNumber.eqv(b: JsNumber): Boolean = when (this) {
+    is JsDecimal -> when (b) {
+      is JsDecimal -> String.eq().run { this@eqv.value.eqv(b.value) }
+      is JsLong -> String.eq().run { this@eqv.value.eqv(b.value.toString()) }
+      is JsDouble -> String.eq().run { this@eqv.value.eqv(b.value.toString()) }
+      is JsFloat -> String.eq().run { this@eqv.value.eqv(b.value.toString()) }
+      is JsInt -> String.eq().run { this@eqv.value.eqv(b.value.toString()) }
+    }
```

Comment:
These decimal comparison looks like a floating point recipe for disaster. The stringify algorithm for floats may yield weird results.

Reply:
<img width="1170" alt="Screen Shot 2019-05-18 at 3 12 43 AM" src="https://user-images.githubusercontent.com/514871/57963346-d586c280-791a-11e9-9ee6-635fc5ab136a.png">